### PR TITLE
[Feature]Support array_contains_seq functions like trino contains_sequence and ck hasSubstr function

### PR DESCRIPTION
**Pull request by leoyy0316**
Fixes #issue
ck:https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hassubstr
trino:https://trino.io/docs/current/functions/array.html?highlight=contains_sequence#contains_sequence
## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/array_functions.cpp [hunk 1]
```diff
@@ -752,6 +752,42 @@ class ArrayHasImpl {
...
+    static uint8 __process_seq(const ElementColumn& elements, uint32 element_start, uint32 element_end,
+                               const ElementColumn& targets, uint32 target_start, uint32 target_end,
+                               const NullColumn::Container* null_map_elements,
+                               const NullColumn::Container* null_map_targets) {
+        using ValueType = std::conditional_t<std::is_same_v<ArrayColumn, ElementColumn> ||
+                                                     std::is_same_v<MapColumn, ElementColumn> ||
+                                                     std::is_same_v<StructColumn, ElementColumn>,
+                                             uint8_t, typename ElementColumn::ValueType>;
+
+        [[maybe_unused]] auto is_null = [](const NullColumn::Container* null_map, size_t idx) -> bool {
```
Comment (LiShuMing) [inline be/src/exprs/array_functions.cpp:766]: remove this if this is not used?

File: be/src/exprs/array_functions.cpp [hunk 2]
```diff
@@ -769,16 +805,23 @@ class ArrayHasImpl {
 
         for (size_t i = 0; i < num_array; i++) {
             uint8_t found = 0;
-            if constexpr (ConstTarget) {
-                DCHECK_EQ(num_target, 1);
-                found = __process<NullableElement, NullableTarget, ElementColumn>(
-                        elements, element_offsets_ptr[i], element_offsets_ptr[i + 1], targets, target_offsets_ptr[0],
-                        target_offsets_ptr[1], null_map_elements, null_map_targets);
-            } else {
+            if (ContainsSeq) {
```
Comment (LiShuMing) [inline be/src/exprs/array_functions.cpp:808]:
```suggestion
            if constexpr (ContainsSeq) {
```

File: test/sql/test_array_fn/R/test_array_fn [hunk 3]
```diff
@@ -4394,4 +4394,64 @@ select array_filter([row(1,2,3), row(3,4,5), row(4,5,6)], [0,1,0]);
...
 -- result:
 2
+-- !result
+select array_contains_seq([1,2,3,4], [2,3]);
+-- result:
+0
+-- !result
+select array_contains_seq([1,2,3,4], [3,2]);
+-- result:
+1
```
Comment (LiShuMing) [inline test/sql/test_array_fn/R/test_array_fn:4404]: error?

File: test/sql/test_array_fn/R/test_array_fn [hunk 4]
```diff
@@ -4394,4 +4394,64 @@ select array_filter([row(1,2,3), row(3,4,5), row(4,5,6)], [0,1,0]);
...
+-- !result
+select array_contains_seq([1,2,3,4], [1,2,3]);
+-- result:
```
Comment (LiShuMing) [inline test/sql/test_array_fn/R/test_array_fn:4407]: error?

**Review by LiShuMing — DISMISSED** [review]
LGTM

File: test/sql/test_array_fn/T/test_array_fn [hunk 5]
```diff
@@ -825,3 +825,19 @@ select array_intersect([row(1,2,3), row(3,4,5)], [row(3,4,5)]);
 select array_contains_all([row(1,2,3), row(3,4,5)], [row(3,4,5)]);
 select array_filter([row(1,2,3), row(3,4,5), row(4,5,6)], [0,1,0]);
 select cardinality([row(1,2,3), row(3,4,5)]);
+
+select array_contains_seq([1,2,3,4], [2,3]);
```
Comment (fzhedu) [inline test/sql/test_array_fn/T/test_array_fn:829]: what if SELECT array_contains_seq([1, 2, NULL, 3, 4], ['a'])

File: docs/sql-reference/sql-functions/array-functions/array_contains_seq.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+The data types of elements in the two arrays must be the same. For the data types of array elements supported by StarRocks, see [ARRAY](../../../sql-reference/sql-statements/data-types/Array.md).
+
+## Return value
+
+Returns a value of the BOOLEAN type.
+
+1 is returned if `arr2` is a subset of `arr1`. Otherwise, 0 is returned.
+
+## Examples
```
Comment (fzhedu) [inline docs/sql-reference/sql-functions/array-functions/array_contains_seq.md:25]: append more special msg like in https://clickhouse.com/docs/en/sql-reference/functions/array-functions#hassubstr, such as null, no supertype etc.

File: test/sql/test_array_fn/T/test_array_fn (hunk 5 above)
Comment (fzhedu) [reply test/sql/test_array_fn/T/test_array_fn:829]: SELECT array_contains_seq([1, 2, NULL, 3, 4], [2,3]);
SELECT array_contains_seq([1, 2, NULL, 3, 4], null);
SELECT array_contains_seq(null, [2,3])
SELECT array_contains_seq([1, 2, NULL, 3, 4], [null,null])
SELECT array_contains_seq([1, 2, NULL], [null,2])
SELECT array_contains_seq(null, null)

Comment (fzhedu) [reply test/sql/test_array_fn/T/test_array_fn:829]: we prefer keep the same behavior with trino

File: docs/sql-reference/sql-functions/array-functions/array_contains_seq.md (hunk 6 above)
Comment (fzhedu) [reply docs/sql-reference/sql-functions/array-functions/array_contains_seq.md:25]: we prefer keeping the same behavior with trino

File: be/src/exprs/array_functions.cpp [hunk 7]
```diff
@@ -752,6 +752,41 @@ class ArrayHasImpl {
...
+            }
+            if (found) {
+                i++;
+                j++;
+            } else {
+                i = 0;
+                j++;
+            }
+        }
+        return i == target_end;
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:788]: seems not process nullable cases?

Comment (leoyy0316) [reply be/src/exprs/array_functions.cpp]: > seems not process nullable cases?

done

File: test/sql/test_array_fn/T/test_array_fn (hunk 5 above)
Comment (leoyy0316) [reply test/sql/test_array_fn/T/test_array_fn:829]: > what if SELECT array_contains_seq([1, 2, NULL, 3, 4], ['a'])

MySQL [(none)]> select array_contains_all([1, 2, NULL, 3, 4], ['a']);
+-----------------------------------------------+
| array_contains_all([1, 2, NULL, 3, 4], ['a']) |
+-----------------------------------------------+
|                                             0 |
+-----------------------------------------------+
1 row in set (0.18 sec)

Comment (leoyy0316) [reply test/sql/test_array_fn/T/test_array_fn:829]: > SELECT array_contains_seq([1, 2, NULL, 3, 4], [2,3]); SELECT array_contains_seq([1, 2, NULL, 3, 4], null); SELECT array_contains_seq(null, [2,3]) SELECT array_contains_seq([1, 2, NULL, 3, 4], [null,null]) SELECT array_contains_seq([1, 2, NULL], [null,2]) SELECT array_contains_seq(null, null)

all add to sqltest

**Review by LiShuMing — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

File: be/src/exprs/array_functions.cpp [hunk 8]
```diff
@@ -752,6 +752,58 @@ class ArrayHasImpl {
...
+        using ValueType = std::conditional_t<std::is_same_v<ArrayColumn, ElementColumn> ||
+                                                     std::is_same_v<MapColumn, ElementColumn> ||
+                                                     std::is_same_v<StructColumn, ElementColumn>,
+                                             uint8_t, typename ElementColumn::ValueType>;
+        [[maybe_unused]] auto is_null = [](const NullColumn::Container* null_map, size_t idx) -> bool {
+            return (*null_map)[idx] != 0;
+        };
+        if (element_end < target_end) {
+            return false;
+        }
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:770]: this is not true, as your tests case just consider single array cases. Taking an array column as example, 
elem: [null][null] (its offsets are [0,0,0]
target: [1,2,3,4],[null] (its offsets are [0,4,4]

array_contains_seq(elem, target) should return false, true. but your answer sould be wrong to false, false.

File: be/src/exprs/array_functions.cpp [hunk 9]
```diff
@@ -752,6 +752,58 @@ class ArrayHasImpl {
...
+                bool null_element = false;
+                if constexpr (NullableElement) {
+                    null_element = is_null(null_map_elements, j);
+                }
+                if (null_target && null_element) {
+                    found = true;
+                } else if (null_target || null_element) {
+                    found = false;
+                } else {
+                    found = (elements_ptr[j] == targets_ptr[i]);
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:794]:
```
            if constexpr (std::is_same_v<ArrayColumn, ElementColumn> || std::is_same_v<MapColumn, ElementColumn> ||
                          std::is_same_v<StructColumn, ElementColumn> || std::is_same_v<JsonColumn, ElementColumn>) {
                found = (elements.equals(j, targets, i) == 1);
            } else
```

should move here

File: test/sql/test_array_fn/T/test_array_fn (hunk 5 above)
Comment (fzhedu) [reply test/sql/test_array_fn/T/test_array_fn:829]: here are all cases that just consider single array, please add some cases for array columns from a table, which would get some bugs.  like following cases

```
CREATE TABLE array_test ( 
pk bigint not null ,
i_0   Array<BigInt>,
i_1   Array<BigInt>,
ai_0  Array<Array<BigInt>>,
ai_1  Array<Array<BigInt>>
) ENGINE=OLAP
DUPLICATE KEY(`pk`)
DISTRIBUTED BY HASH(`pk`) BUCKETS 3
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false"
);

insert into array_test values
(1,[null,1],[null],[null],[[]]),
(2,[null],[1,3,4,5,1,2],[[null,1],null],[[1,null]]),
(3,[],[],[[],null,[1,1]],[],[1,1]),
(4,null,[],[[1,1]],[[1,1],null),
(5,[4,4,4],[4,null],[null],[null]),
(6,[1,1,2,1,1,2,3,3],[1,2,3],[[1]],[[1],[2],null,[null]]);

select array_contains_seq(i_0,i_1),array_contains_seq(i_0,i_0),array_contains_seq(i_1,i_0), array_contains_seq(ai_0, ai_1),array_contains_seq(ai_1, ai_1),array_contains_seq(ai_1, ai_0),array_contains_seq(ai_0, i_1) from array_test;

```

**Review by fzhedu — COMMENTED** [review]
please refer to be/test/exprs/array_functions_test.cpp to write some unit tests.
array_contains_empty_array
array_contains_all

File: be/src/exprs/array_functions.cpp (hunk 8 above)
Comment (fzhedu) [reply be/src/exprs/array_functions.cpp:770]: if(element_end-element_start > target_end-target_start)

File: be/src/exprs/array_functions.cpp [hunk 10]
```diff
@@ -752,6 +752,73 @@ class ArrayHasImpl {
...
+        }
+        if (element_end == element_start) {
+            return false;
+        }
+        bool found = false;
+        size_t i = target_start;
+        size_t j = element_start;
+        while (j < element_end) {
+            int k = j;
+            int l = 0;
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:782]: remove k,l;
i = target_start;

File: be/src/exprs/array_functions.cpp [hunk 11]
```diff
@@ -752,6 +752,73 @@ class ArrayHasImpl {
...
+                i++;
+                if (found) {
+                    k++;
+                    l++;
+                }
+            }
+            if (l == target_end) {
+                return true;
+            }
+            i = 0;
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:817]: remove

File: be/src/exprs/array_functions.cpp [hunk 12]
```diff
@@ -752,6 +752,73 @@ class ArrayHasImpl {
...
+                        auto elements_ptr = (const ValueType*)(elements.raw_data());
+                        auto targets_ptr = (const ValueType*)(targets.raw_data());
+                        found = (elements_ptr[k] == targets_ptr[i]);
+                    }
+                }
+                i++;
+                if (found) {
+                    k++;
+                    l++;
+                }
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:812]: else break;

File: be/test/exprs/array_functions_test.cpp [hunk 13]
```diff
@@ -5467,4 +5467,55 @@ TEST_F(ArrayFunctionsTest, array_match_only_null) {
...
+    // array_contains_seq(["a", "b", "c"], ["c"])         -> 1
+    // array_contains_seq(NULL, ["c"])                    -> NULL
+    // array_contains_seq(["a", "b", "c"], NULL)          -> NULL
+    // array_contains_seq(["a", "b", NULL], NULL)         -> NULL
+    // array_contains_seq(["a", "b", NULL], ["a", NULL])  -> 0
+    // array_contains_seq(NULL, ["a", NULL])              -> NULL
+    // array_contains_seq(["a", "b", NULL], [NULL])       -> 1
+    // array_contains_seq(["a", "b", "c"], ["d"])         -> 0
+    // array_contains_seq(["a", "b", "c"], ["a", "d"])    -> 0
+    // array_contains_all(["a", "b", "c"], ["a", "c"])    -> 1
```
Comment (fzhedu) [inline be/test/exprs/array_functions_test.cpp:5481]: ??

File: be/test/exprs/array_functions_test.cpp [hunk 14]
```diff
@@ -5467,4 +5467,55 @@ TEST_F(ArrayFunctionsTest, array_match_only_null) {
         ASSERT_TRUE(dest_column->get(0).get_int8());
     }
 }
+// NOLINTNEXTLINE
+TEST_F(ArrayFunctionsTest, array_contains_seq) {
```
Comment (fzhedu) [inline be/test/exprs/array_functions_test.cpp:5471]: here is the case nullable array1 and nullable array2, pls refer to array_contains_no_null to add some cases from nullable array1, not-nullable array2, not-nullable array1, nullable array2, not_nullable array1, not_nullable array2.

File: test/sql/test_array_fn/T/test_array_fn (hunk 5 above)
Comment (fzhedu) [reply test/sql/test_array_fn/T/test_array_fn:829]: add this case to the Test file

File: be/src/exprs/array_functions.cpp [hunk 15]
```diff
@@ -752,6 +752,73 @@ class ArrayHasImpl {
...
+        if (target_end == target_start) {
+            return true;
+        }
+        if (element_end == element_start) {
+            return false;
+        }
+        bool found = false;
+        size_t i = target_start;
+        size_t j = element_start;
+        while (j < element_end) {
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:780]: j <= element_end - (target_end - target_start), no need to loop if the left elements are less than target's elements

File: be/test/exprs/array_functions_test.cpp [hunk 16]
```diff
@@ -5467,4 +5467,116 @@ TEST_F(ArrayFunctionsTest, array_match_only_null) {
...
+        EXPECT_EQ(1, result->get(0).get_int8());
+        EXPECT_EQ(1, result->get(2).get_int8());
+    }
+    // array_contains_seq([["a"], ["b"], [NULL]], [["c"]])
+    // array_contains_seq(["a","d","c"], ["e"])
+    // array_contains_seq([["a", "b"], ["c"]], [["a", "b"]])
+    {
+        auto array = ColumnHelper::create_column(TYPE_ARRAY_ARRAY_VARCHAR, true);
+        array->append_datum(DatumArray{Datum(DatumArray{"a"}), Datum(DatumArray{"b"}), Datum()});
+        array->append_datum(DatumArray{"a", "d", "c"});
```
Comment (fzhedu) [inline be/test/exprs/array_functions_test.cpp:5547]: should be [[xxx]]

File: be/src/exprs/array_functions.cpp [hunk 17]
```diff
@@ -752,6 +752,76 @@ class ArrayHasImpl {
...
+                } else {
+                    break;
+                }
+            }
+            if (i == target_end) {
+                return true;
+            }
+            if ((element_end - j) < (target_end - target_start)) {
+                return false;
+            }
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:820]: if put here, it should be <=; prefer put this condition to the while() at L:780

File: test/sql/test_array_fn/T/test_array_fn [hunk 18]
```diff
@@ -825,3 +825,27 @@ select array_intersect([row(1,2,3), row(3,4,5)], [row(3,4,5)]);
...
+select array_contains_seq([json_keys('{"a":1,"b":2}')], [json_keys('{"a":1,"b":2}')]);
+select array_contains_seq([map([1,2,5],[2,4,5])], [map([1,2,5],[2,4,5])]);
+select array_contains_seq([map([1,2,5],[2,4,5])], [map([1,2],[2,4])]);
+select array_contains_seq([1, 2, NULL, 3, 4], ['a']);
+select array_contains_seq([1, 2, NULL, 3, 4], [2,3]);
+select array_contains_seq([1, 2, NULL, 3, 4], null);
+select array_contains_seq(null, [2,3]);
+select array_contains_seq([1, 2, NULL, 3, 4], [null,null]);
+select array_contains_seq([1, 2, NULL], [null,2]);
+select array_contains_seq(null, null);
```
Comment (fzhedu) [inline test/sql/test_array_fn/T/test_array_fn:850]: add these tests to this file
```
CREATE TABLE array_test ( pk bigint not null , i_0 Array<BigInt>, i_1 Array<BigInt>, ai_0 Array<Array<BigInt>>, ai_1 Array<Array<BigInt>> ) ENGINE=OLAP DUPLICATE KEY(`pk`) DISTRIBUTED BY HASH(`pk`) BUCKETS 3 PROPERTIES ( "replication_num" = "1", "in_memory" = "false" ); insert into array_test values (1,[null,1],[null],[null],[[]]), (2,[null],[1,3,4,5,1,2],[[null,1],null],[[1,null]]), (3,[],[],[[],null,[1,1]],[],[1,1]), (4,null,[],[[1,1]],[[1,1],null), (5,[4,4,4],[4,null],[null],[null]), (6,[1,1,2,1,1,2,3,3],[1,2,3],[[1]],[[1],[2],null,[null]]); select array_contains_seq(i_0,i_1),array_contains_seq(i_0,i_0),array_contains_seq(i_1,i_0), array_contains_seq(ai_0, ai_1),array_contains_seq(ai_1, ai_1),array_contains_seq(ai_1, ai_0),array_contains_seq(ai_0, i_1),array_contains_seq(ai_0,null),,array_contains_seq(ai_0, [1,2]) from array_test;
```

**Review by fzhedu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**